src/main/clojure/clara/rules/compiler.clj

```diff
@@ -415,17 +415,20 @@
         (list `-> '?__token__ :bindings binding-key)))
 
 ;; FIXME: add env...
-(defn compile-test [node-id tests]
-  (let [binding-keys (variables-as-keywords tests)
+(defn compile-test [node-id constraints]
+  (let [binding-keys (variables-as-keywords constraints)
         assignments (mapcat build-token-assignment binding-keys)
 
         ;; Hardcoding the node-type and fn-type as we would only ever expect 'compile-test' to be used for this scenario
         fn-name (mk-node-fn-name "TestNode" node-id "TE")]
 
-    `(fn ~fn-name [~'?__token__]
-       (let [~@assignments]
-
-        (and ~@tests)))))
+    `(let [handler# (fn ~fn-name [~'?__token__]
+                      (let [~@assignments]
+                        (and ~@constraints)))]
+       (with-meta handler#
```

Comment:
There may be a reason to be concerned here about the overhead of using `with-meta` to wrap the fn.

`with-meta` wraps the fn in a RestFn here https://github.com/clojure/clojure/blob/clojure-1.10.3/src/jvm/clojure/lang/AFunction.java#L29
which has the unfortunate consequence of making all calls to the wrapped fn go through variadic arg runtime unrolling via `apply` https://github.com/clojure/clojure/blob/clojure-1.10.3/src/jvm/clojure/lang/AFunction.java#L31

The constraint fn's themselves may be called in hot spots so this can be concerning.

Microbench mark example (ran it a bunch of times to stabilize the numbers):
```clj
(let [f (fn [x] (inc x))
      mf (with-meta f {:my :meta})]

  (time (dotimes [_ 1e5]
          (f 2)))

  (time (dotimes [_ 1e5]
          (mf 2))))
;;= "Elapsed time: 3.851125 msecs"
;;= "Elapsed time: 22.38039 msecs"
;;= nil
```

Reply:
True, there’s a bit of an overhead, the alternative approach I was considering was creating a protocol for the test expression handler instead of using a Fn directly so that it could carry the additional metadata needed for introspection at runtime, or just simply capturing that data when compile-test is called. It will require refactoring more tests but I can make an alternative PR with that approach.

Reply:
One possible way out of this sort of mess would be if this fn could just return this map instead and then 
https://github.com/cerner/clara-rules/blob/014ed43ed2960d64db065a1fd108dce336aac960/src/main/clojure/clara/rules/compiler.clj#L1443
would use the `:handler` value directly.

However, I think then there may be an issue with the durability layer persisting this meta information via serde...

Reply:
@mrrodriguez in my experience it gets persisted just fine because the durability layer persists the productions but removes the compiled portion, so it would recompile them whenever the session was reassembled.